GitUI/CommandsDialogs/FormBrowse.cs

```diff
@@ -3077,154 +3072,215 @@ private void LoadSubmodulesIntoDropDownMenu()
+                    info.Text += submoduleStatus.AddedAndRemovedString();
+                }
+            }, cancelToken, TaskCreationOptions.AttachedToParent, TaskScheduler.Default);
         }
 
         private void UpdateSubmodulesList()
         {
-            RemoveSubmoduleButtons();
             _previousUpdateTime = DateTime.Now;
-            _submodulesStatusImagesCTS = new CancellationTokenSource();
-
-            foreach (var submodule in Module.GetSubmodulesLocalPathes().OrderBy(submoduleName => submoduleName))
-            {
-                var name = submodule;
-                string path = Module.GetSubmoduleFullPath(submodule);
-                if (Settings.DashboardShowCurrentBranch && !GitModule.IsBareRepository(path))
-                    name = name + " " + GetModuleBranch(path);
 
-                var smi = AddSubmoduleToMenu(name, path);
-                var module = Module.GetSubmodule(submodule);
-                var submoduleName = module.GetCurrentSubmoduleLocalPath();
-                GetSubmoduleStatusImageAsync(smi, module.SuperprojectModule, submoduleName);
-            }
+            // Cancel any previous async activities:
+            _submodulesStatusCTS.Cancel();
+            _submodulesStatusCTS.Dispose();
+            _submodulesStatusCTS = new CancellationTokenSource();
 
-            bool containSubmodules = toolStripButtonLevelUp.DropDownItems.Count != 0;
-            if (!containSubmodules)
-                toolStripButtonLevelUp.DropDownItems.Add(_noSubmodulesPresent.Text);
+            RemoveSubmoduleButtons();
+            toolStripButtonLevelUp.DropDownItems.Add(_loading.Text);
 
-            string currentSubmoduleName = null;
-            if (Module.SuperprojectModule != null)
+            // Start gathering new submodule information asynchronously.  This makes a significant difference in UI
+            // responsiveness if there are numerous submodules (e.g. > 100).
+            var cancelToken = _submodulesStatusCTS.Token;
+            string thisModuleDir = Module.WorkingDir;
+            // First task: Gather list of submodules on a background thread.
+            var updateTask = Task.Factory.StartNew(() =>
             {
-                var superprojectSeparator = new ToolStripSeparator();
-                toolStripButtonLevelUp.DropDownItems.Add(superprojectSeparator);
+                // Don't access Module directly because it's not thread-safe.  Use a thread-local version:
+                GitModule threadModule = new GitModule(thisModuleDir);
+                SubmoduleInfoResult result = new SubmoduleInfoResult();
 
-                GitModule supersuperproject = Module.FindTopProjectModule();
-                if (Module.SuperprojectModule.WorkingDir != supersuperproject.WorkingDir)
+                // Add all submodules inside the current repository:
+                foreach (var submodule in threadModule.GetSubmodulesLocalPathes().OrderBy(submoduleName => submoduleName))
```

Comment:
This was a problem in the original code. I'll add a separate commit to fix this.